### PR TITLE
Update create commands

### PR DESCRIPTION
**Pull request by juanvallejo**
Updates create commands with patterns established upstream.

cc @soltysh @deads2k

### CODE REVIEW

File: pkg/cmd/util/route/route.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ import (
 // forcePort always sets a port, even when there is only one and it has no name.
 // The kubernetes generator, when no port is present incorrectly selects the service Port
 // instead of the service TargetPort for the route TargetPort.
-func UnsecuredRoute(kc kclientset.Interface, namespace, routeName, serviceName, portString string, forcePort bool) (*routeapi.Route, error) {
+func UnsecuredRoute(kc kclientset.Interface, namespace, routeName, serviceName, portString string, forcePort bool) (*routev1.Route, error) {
```
Comment (deads2k) [inline pkg/cmd/util/route/route.go]: move entire package to `pkg/oc` somewhere

File: pkg/oc/cli/cmd/create/clusterquota.go [hunk 2]
```diff
@@ -35,79 +33,87 @@ var (
 )
 
 type CreateClusterQuotaOptions struct {
-	ClusterQuota *quotaapi.ClusterResourceQuota
-	Client       quotaclient.ClusterResourceQuotasGetter
+	*CreateSubcommandOptions
```
Comment (deads2k) [inline pkg/oc/cli/cmd/create/clusterquota.go]: strongly dislike anonymous inclusion.  Why is it necessary here?

File: pkg/oc/cli/cmd/create/clusterquota.go [hunk 3]
```diff
@@ -35,79 +33,87 @@ var (
...
 
-	DryRun bool
+	LabelSelectorStr      string
```
Comment (deads2k) [inline pkg/oc/cli/cmd/create/clusterquota.go]: ProjectLabelSelectorStr

File: pkg/oc/cli/cmd/create/clusterquota.go [hunk 4]
```diff
@@ -35,79 +33,87 @@ var (
...
+	AnnotationSelectorStr string
```
Comment (deads2k) [inline pkg/oc/cli/cmd/create/clusterquota.go]: ProjectAnnotationSelectorStr

File: pkg/oc/cli/cmd/create/create.go [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	PrintFlags *create.PrintFlags
+	// Name of resource being created
+	Name string
+	// DryRun is true if the command should be simulated but not run against the server
+	DryRun bool
+
+	Namespace        string
+	EnforceNamespace bool
+
+	Mapper meta.RESTMapper
```
Comment (deads2k) [inline pkg/oc/cli/cmd/create/create.go]: unexpected.  Why does a create command need a restmapper?

Comment (deads2k) [reply pkg/oc/cli/cmd/create/create.go]: it doesn't look generic at any rate

File: pkg/oc/cli/cmd/create/deploymentconfig.go [hunk 6]
```diff
@@ -32,131 +30,87 @@ var (
...
-	Out          io.Writer
-	Printer      ObjectPrinter
+	Client appsv1client.DeploymentConfigsGetter
 }
 
 // NewCmdCreateDeploymentConfig is a macro command to create a new deployment config.
-func NewCmdCreateDeploymentConfig(name, fullName string, f kcmdutil.Factory, out io.Writer) *cobra.Command {
-	o := &CreateDeploymentConfigOptions{Out: out}
-
+func NewCmdCreateDeploymentConfig(name, fullName string, f kcmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
```
Comment (deads2k) [inline pkg/oc/cli/cmd/create/deploymentconfig.go]: does it need a factory or just the restclientgetter?  narrow scope where easily possible.

File: pkg/oc/cli/cmd/create/deploymentconfig.go [hunk 7]
```diff
@@ -2,20 +2,18 @@ package create
 
 import (
 	"fmt"
-	"io"
 
 	"github.com/spf13/cobra"
 
+	kapiv1 "k8s.io/api/core/v1"
```
Comment (deads2k) [inline pkg/oc/cli/cmd/create/deploymentconfig.go]: use corev1

File: pkg/oc/cli/cmd/create/imagestreamtag.go [hunk 8]
```diff
@@ -42,105 +41,110 @@ var (
...
 		Run: func(cmd *cobra.Command, args []string) {
 			cmdutil.CheckErr(o.Complete(cmd, f, args))
-			cmdutil.CheckErr(o.Validate())
 			cmdutil.CheckErr(o.Run())
 		},
 		Aliases: []string{"istag"},
 	}
+	cmd.Flags().StringVar(&o.FromImage, "from-image", o.FromImage, "Use the provided remote image with this tag.")
+	cmd.Flags().StringVar(&o.From, "from", o.From, "Use the provided image stream tag or image stream image as the source: [<namespace>/]name[:<tag>|@<id>]")
+	cmd.Flags().StringSliceVarP(&o.Annotations, "annotation", "A", o.Annotations, "Set an annotation on this image stream tag.")
```
Comment (deads2k) [inline pkg/oc/cli/cmd/create/imagestreamtag.go]: oh dear.  I wonder how this got in.

Comment (juanvallejo) [reply pkg/oc/cli/cmd/create/imagestreamtag.go]: Will add deprecation message for that shorthand

**Comment by juanvallejo** [comment]
@deads2k thanks, review comments addressed. Will squash if everything looks okay

File: pkg/oc/cli/cmd/create/user_identity_mapping.go [hunk 9]
```diff
@@ -83,12 +83,6 @@ func (o *CreateUserIdentityMappingOptions) Complete(cmd *cobra.Command, f kcmdut
 		return err
 	}
 
-	// this is copied from CreateSubcommandOptions.Complete because this command
-	// parses incoming arguments differently
-	o.CreateSubcommandOptions.Mapper, err = f.ToRESTMapper()
```
[the remark this answers is not in the document]
Comment (soltysh) [reply pkg/oc/cli/cmd/create/user_identity_mapping.go]: Only that I've seen it elsewhere, but I haven't checked if it was used. If it's not - kill it!

File: pkg/oc/cli/cmd/create/clusterquota.go [hunk 10]
```diff
@@ -35,79 +32,87 @@ var (
...
-type ObjectPrinter func(runtime.Object, io.Writer) error
+	Client quotav1client.ClusterResourceQuotasGetter
+}
 
 // NewCmdCreateClusterQuota is a macro command to create a new cluster quota.
-func NewCmdCreateClusterQuota(name, fullName string, f kcmdutil.Factory, out io.Writer) *cobra.Command {
-	o := &CreateClusterQuotaOptions{Out: out}
-
+func NewCmdCreateClusterQuota(name, fullName string, f genericclioptions.RESTClientGetter, streams genericclioptions.IOStreams) *cobra.Command {
+	o := &CreateClusterQuotaOptions{
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/clusterquota.go]: I didn't do it before, but let's create `NewCreateClusterQuotaOptions` method like we do in all other places. It's nice and clean solution.

File: pkg/oc/cli/cmd/create/route.go [hunk 11]
```diff
@@ -27,423 +26,90 @@ var (
...
-		return err
-	}
-
-	wildcardpolicy := kcmdutil.GetFlagString(cmd, "wildcard-policy")
-	if len(wildcardpolicy) > 0 {
-		route.Spec.WildcardPolicy = routeapi.WildcardPolicyType(wildcardpolicy)
-	}
+// CreateRouteSubcommandOptions is an options struct to support create subcommands
+type CreateRouteSubcommandOptions struct {
+	genericclioptions.IOStreams
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/route.go]: At the bottom.

File: pkg/oc/cli/cmd/create/user_identity_mapping.go [hunk 12]
```diff
@@ -2,19 +2,17 @@ package create
 
 import (
 	"fmt"
-	"io"
 
 	"github.com/spf13/cobra"
 
-	"k8s.io/apimachinery/pkg/runtime"
-	kapi "k8s.io/kubernetes/pkg/apis/core"
+	kapiv1 "k8s.io/api/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/user_identity_mapping.go]: `corev1`

File: pkg/oc/cli/cmd/create/clusterquota.go [hunk 13]
```diff
@@ -3,22 +3,19 @@ package create
 import (
 	"encoding/csv"
 	"fmt"
-	"io"
 	"strings"
 
 	"github.com/spf13/cobra"
 
+	kapiv1 "k8s.io/api/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/clusterquota.go]: `corev1`

File: pkg/oc/cli/cmd/create/imagestreamtag.go [hunk 14]
```diff
@@ -2,21 +2,19 @@ package create
 
 import (
 	"fmt"
-	"io"
 	"strings"
 
 	"github.com/spf13/cobra"
 
+	kapiv1 "k8s.io/api/core/v1"
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/imagestreamtag.go]: `corev1`

**Comment by juanvallejo** [comment]
@soltysh thanks, comments addressed

File: pkg/oc/cli/cmd/create/create.go [hunk 15]
```diff
@@ -0,0 +1,69 @@
...
+	"k8s.io/kubernetes/pkg/api/legacyscheme"
+	"k8s.io/kubernetes/pkg/kubectl/cmd/create"
+	cmdutil "k8s.io/kubernetes/pkg/kubectl/cmd/util"
+	"k8s.io/kubernetes/pkg/kubectl/genericclioptions"
+	"k8s.io/kubernetes/pkg/kubectl/genericclioptions/printers"
+)
+
+// CreateSubcommandOptions is an options struct to support create subcommands
+type CreateSubcommandOptions struct {
+	genericclioptions.IOStreams
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/create.go]: At the bottom.

File: pkg/oc/cli/cmd/create/routeedge.go [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+	Path           string
+	Cert           string
+	Key            string
+	CACert         string
+	WildcardPolicy string
+}
+
+// NewCmdCreateEdgeRoute is a macro command to create an edge route.
+func NewCmdCreateEdgeRoute(fullName string, f kcmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
+	o := &CreateEdgeRouteOptions{
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/routeedge.go]: Turn that into a method like we do with all the other `*Options`.

File: pkg/oc/cli/cmd/create/routepassthrough.go [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+	Hostname       string
+	Port           string
+	InsecurePolicy string
+	Service        string
+	WildcardPolicy string
+}
+
+// NewCmdCreatePassthroughRoute is a macro command to create a passthrough route.
+func NewCmdCreatePassthroughRoute(fullName string, f kcmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
+	o := &CreatePassthroughRouteOptions{
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/routepassthrough.go]: Turn that into a method like we do with all the other `*Options`.

File: pkg/oc/cli/cmd/create/routereenecrypt.go [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+	Cert           string
+	Key            string
+	CACert         string
+	DestCACert     string
+	WildcardPolicy string
+}
+
+// NewCmdCreateReencryptRoute is a macro command to create a reencrypt route.
+func NewCmdCreateReencryptRoute(fullName string, f kcmdutil.Factory, streams genericclioptions.IOStreams) *cobra.Command {
+	o := &CreateReencryptRouteOptions{
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/routereenecrypt.go]: Turn that into a method like we do with all the other `*Options`.

File: pkg/oc/cli/cmd/create/user.go [hunk 19]
```diff
@@ -35,107 +33,65 @@ var (
...
-	Printer      ObjectPrinter
+	UserClient userv1client.UsersGetter
 }
 
 // NewCmdCreateUser is a macro command to create a new user
-func NewCmdCreateUser(name, fullName string, f kcmdutil.Factory, out io.Writer) *cobra.Command {
-	o := &CreateUserOptions{Out: out}
-
+func NewCmdCreateUser(name, fullName string, f genericclioptions.RESTClientGetter, streams genericclioptions.IOStreams) *cobra.Command {
+	o := &CreateUserOptions{
```
Comment (soltysh) [inline pkg/oc/cli/cmd/create/user.go]: Turn that into a method like we do with all the other `*Options`.

**Pull request merged**